### DISCUSSION TITLE
Get message_id when sending sms

### DISCUSSION THREAD

**Discussion by micartey** [python-smpplib/python-smpplib#246] [Q&A] [answered]
The example contains the following snipped for sending a SMS:

```python
for part in parts:
    pdu = client.send_message(
        source_addr_ton=smpplib.consts.SMPP_TON_INTL,
        #source_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
        # Make sure it is a byte string, not unicode:
        source_addr='SENDERPHONENUM',

        dest_addr_ton=smpplib.consts.SMPP_TON_INTL,
        #dest_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
        # Make sure thease two params are byte strings, not unicode:
        destination_addr='PHONENUMBER',
        short_message=part,

        data_coding=encoding_flag,
        esm_class=msg_type_flag,
        registered_delivery=True,
    )
    print(pdu.sequence)
```

Is it somehow possible to get the `message_id` that will be later echoed as `receipted_message_id` at this point?
I know that I can use the following listener:

```python
client.set_message_sent_handler(
    lambda pdu: logger.debug(f"sent {pdu.message_id}")
)
```

But I cannot know if the message_id is the message id for the message I just sent in a multi-threaded environment 🤔

**Comment by eigenein** [comment]
👋 Message ID indeed comes back as a part of `deliver_sm_resp` so it is not possible to obtain it directly from `client.send_message()`. My guess would be to call `client.read_pdu()` immediately after `client.send_message()` to receive that `deliver_sm_resp` and read message ID from there

**Comment by micartey** [comment]
I tried calling `client.read_pdu()`, but that didn't work.
It didn't do anything and after some time I got the following exception:

```
smpplib.exceptions.UnknownCommandError: Unknown SMPP command code '0x0'
```

Even if this works, the delay is a huge issue.
Especially for messages that consist of different parts.

**Comment by eigenein** [comment]
For the `UnknownCommandError` we have the issue #105 and there are no hands to implement that at the moment (although, I'd be happy to review a PR).

> I tried calling `client_read_pdu()`, but that didn't work.
> It didn't do anything

How do you mean «didn't do anything»? Like being stuck?

**Comment by micartey** [comment]
Yes, it was stuck until the exception was thrown. Therefore it also blocked the other parts when I chose a longer message

**Comment by micartey** [comment, marked as answer]
I resolved the issue by entering a lock when sending an SMS and releasing it when all parts can be associated with a `message_id` provided by the `set_message_sent_handler`. The time we are waiting is neglectable as the handler is usually called before the second part of the message is processed.
